### PR TITLE
fix: sanitize html blocks

### PR DESCRIPTION
**Pull request by kellyjosephprice**
[![PR App][icn]][demo] |
:-------------------:|

## 🧰 Changes

We were stripping `script` tags from html magic blocks, but that doesn't prevent executing javascript from attributes. This runs the html through the html converter which includes a sanitization step.

Examples:

```
[block:html]
{
  "html": "<body onload=\"alert('wowzas')\"></body>"
}
[/block]
```

```
[block:html]
{
  "html": "<style>@keyframes x{}</style><xss style=\"animation-name:x\" onanimationstart=\"alert(document.cookie)\"></xss>"
}
[/block]
```

## 🧬 QA & Testing

- [Broken on production][prod].
- [Working in this PR app][demo].

[demo]: https://markdown-pr-PR_NUMBER.herokuapp.com
[prod]: https://SUBDOMAIN.readme.io
[icn]: https://user-images.githubusercontent.com/886627/160426047-1bee9488-305a-4145-bb2b-09d8b757d38a.svg

BREAKING CHANGE: This strips some tags and attibutes from html blocks. One such tag that is no longer allowed is `button`.

### CODE REVIEW

File: processor/parse/magic-block-parser.js [hunk 1]
```diff
@@ -247,7 +248,7 @@ function tokenize(eat, value) {
             data: {
               hName: 'html-block',
               hProperties: {
-                html: json.html,
+                html: parser.transforms.html(json.html),
```
Comment (erunion) [inline processor/parse/magic-block-parser.js:251]: Do we need to update non-magic block Markdown parsing too to exclude these unsafe attributes?

Comment (kellyjosephprice) [reply processor/parse/magic-block-parser.js:251]: The sanitizing happens for all the transformers, it's just not smart enough to know about magic blocks: 
https://github.com/readmeio/markdown/blob/next/index.js#L124

Comment (erunion) [reply processor/parse/magic-block-parser.js:251]: mind adding a test for regular markdown for these cases just to be safe?

File: __tests__/magic-block-parser/html.test.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      <React.Fragment>
+        <HTMLBlock
+          html="<div>Uh oh.</div>"
+          runScripts={false}
+        />
+      </React.Fragment>
+    `);
+  });
+
+  it('does not allow arbitrary markdown', () => {
```
Comment (kellyjosephprice) [inline __tests__/magic-block-parser/html.test.js:50]: Just realized the first pass was allowing arbitrary markdown in html blocks...:grimacing:

File: processor/parse/magic-block-parser.js [hunk 3]
```diff
@@ -1,3 +1,11 @@
+const unified = require('unified');
+const rehypeParse = require('rehype-parse');
+const rehypeSanitize = require('rehype-sanitize');
+const rehypeStringify = require('rehype-stringify');
+
+const globalSanitizeSchema = require('../../sanitize.schema');
+
+// eslint-disable-next-line eslint-comments/disable-enable-pair
```
Comment (erunion) [inline processor/parse/magic-block-parser.js:8]: Can maybe remove this?

**Review by erunion — APPROVED** [review]

**Comment by rafegoldberg** [comment]
## This PR was released!

🚀 Changes included in [<b>v6.41.1</b>](/readmeio/markdown/releases/tag/v6.41.1)
<br>

**Review by domharrington — COMMENTED** [review]
Thanks for jumping on this! 👍 Code LGTM

File: __tests__/magic-block-parser.test.js [hunk 4]
```diff
@@ -240,9 +240,10 @@ describe('Parse Magic Blocks', () => {
   it('Custom HTML Block', () => {
     const text = `[block:html]
     ${JSON.stringify({
-      html: '<h1>👋🌍</h1>\n<hr>\n<form>\n  <input name="test" value="hello" type="text"/>\n  <br/>\n  <a class="button">Go</a>\n</form>',
```
Comment (rafegoldberg) [inline __tests__/magic-block-parser.test.js:243]: Our users probably have inputs and forms in these custom HTML block which they'll expect to work… 😬

**Review by rafegoldberg — COMMENTED** [review]
I probably wouldn't've fixed this issue with the sanitizer as I think it's quite likely to cause headaches for customers. I think a likely less breakable route would have been to simply have disabled the custom HTML block in suggested edits in the ReadMe side (somehow…)

**Pull request merged**